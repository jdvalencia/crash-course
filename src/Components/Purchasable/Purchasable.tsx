
import React, { FunctionComponent, ReactElement } from 'react';

export interface IPurchasableProps {
    purchased:boolean
    render:(props:any)=>ReactElement
    price:number
    buyPage:()=>void
    cookies:number
}

//higher order component
export const Purchasable:FunctionComponent<IPurchasableProps> = (props) => {


    return (

        props.purchased ?
        props.render(props)//if we needed to pass in data fom the purchasable component we could easily
        :
        <div>
            <p>You need to purchase this page</p>
            <p>It costs {props.price} Cookies</p>
            <button disabled={props.cookies < props.price} onClick={props.buyPage}>Buy Now</button>
        </div>
    )
}