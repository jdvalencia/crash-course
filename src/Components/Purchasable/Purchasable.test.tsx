import React from 'react';
import { shallow } from 'enzyme';
import {IPurchasableProps, Purchasable} from './Purchasable';


describe('<Purchasable> with no props', ()=> {

    it('should match previous snapshot?', ()=> {
        //@ts-ignore
        let wrapper = shallow(<Purchasable/>)
        expect(wrapper.html()).toMatchSnapshot();
    })
})

const propsUtil = (cookies: number, price: number, purchased:boolean):IPurchasableProps=> {
    let props = {

        render: jest.fn((()=><div>rendered</div>)),
        buyPage: jest.fn(()=> {
            props.purchased = true;
            props.cookies -= price}),
        cookies, 
        price, 
        purchased
    }
    return props;
}

describe('<Purchasable> with props', ()=>{

    it('should show purchase html when purchased is false', ()=>{
        let props = propsUtil(0,0,false)
        let wrapper = shallow(<Purchasable {...props} />)
        expect(wrapper.length).toBe(1)
        expect(wrapper.children().length).toBe(3)
        expect(wrapper.children().contains(<button disabled={props.cookies < props.price} onClick={props.buyPage}>Buy Now</button>)).toBeTruthy()
    })

    it('should call props.render if purchased is true', ()=>{
        let props = propsUtil(0,0,true)
        let wrapper = shallow(<Purchasable {...props} />)
        expect(props.render).toBeCalled()
        expect(wrapper.text()).toMatch('rendered')
    })

})