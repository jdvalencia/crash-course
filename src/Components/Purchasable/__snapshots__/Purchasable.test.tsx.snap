// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Purchasable> with no props should match previous snapshot? 1`] = `"<div><p>You need to purchase this page</p><p>It costs  Cookies</p><button>Buy Now</button></div>"`;
